src/box/space.c

```diff
@@ -144,7 +144,8 @@ space_init_constraints(struct space *space)
 		if (constr->check != tuple_constraint_noop_check)
 			continue;
 		if (constr->def.type == CONSTR_FUNC) {
-			if (tuple_constraint_func_init(constr, space) != 0)
+			if (tuple_constraint_func_init(constr, space,
+						       true) != 0)
```

Comment:
Sorry, I don't see how adding SQL headers would help here. The only reason I added this flag is to find out if this constraint is a field constraint. We can think of another way, but I think adding a third argument is the simplest way. Also the `tuple_constraint_fkey_init()` function has a similar third argument, so I figured that it wouldn't be a problem. I asked Vova, and he said that it was fine for him.

I think it's better to leave this function in `sql/func.c` since it works with the internals of `struct Vdbe`.

Reply:
Ok, if you don't care about headers, then it is even simpler to reduce the complexity:
<details><summary>Diff</summary>

```Diff
diff --git a/src/box/space.c b/src/box/space.c
index 73c8afc34..b0df79c81 100644
--- a/src/box/space.c
+++ b/src/box/space.c
@@ -54,6 +54,7 @@
 #include "tuple_constraint_func.h"
 #include "tuple_constraint_fkey.h"
 #include "wal_ext.h"
+#include "sql/func.h"
 
 int
 access_check_space(struct space *space, user_access_t access)
@@ -144,8 +145,7 @@ space_init_constraints(struct space *space)
 		if (constr->check != tuple_constraint_noop_check)
 			continue;
 		if (constr->def.type == CONSTR_FUNC) {
-			if (tuple_constraint_func_init(constr, space,
-						       false) != 0)
+			if (tuple_constraint_func_init(constr, space) != 0)
 				return -1;
 		} else {
 			assert(constr->def.type == CONSTR_FKEY);
@@ -161,17 +161,27 @@ space_init_constraints(struct space *space)
 					   constr->def.type == CONSTR_FKEY;
 			if (constr->check != tuple_constraint_noop_check)
 				continue;
-			if (constr->def.type == CONSTR_FUNC) {
-				if (tuple_constraint_func_init(constr,
-							       space,
-							       true) != 0)
-					return -1;
-			} else {
-				assert(constr->def.type == CONSTR_FKEY);
+			if (constr->def.type == CONSTR_FKEY) {
 				if (tuple_constraint_fkey_init(constr,
 							       space, i) != 0)
 					return -1;
+				continue;
 			}
+			assert(constr->def.type == CONSTR_FUNC);
+			if (tuple_constraint_func_init(constr, space) != 0)
+				return -1;
+			const struct func *func =
+				constr->func_cache_holder.func;
+			if (func->def->language != FUNC_LANGUAGE_SQL_EXPR)
+				continue;
+			if (func_sql_expr_has_single_arg(func))
+				continue;
+			diag_set(ClientError, ER_CREATE_CONSTRAINT,
+				 constr->def.name, space->def->name, "Number "
+				 "of arguments in a SQL field constraint "
+				 "function is greater than one");
+			constr->space = NULL;
+			return -1;
 		}
 	}
 	space->has_foreign_keys = has_foreign_keys;
diff --git a/src/box/sql.h b/src/box/sql.h
index 7a0868a82..932677ca3 100644
--- a/src/box/sql.h
+++ b/src/box/sql.h
@@ -429,10 +429,6 @@ vdbe_field_ref_prepare_array(struct vdbe_field_ref *ref, uint32_t field_count,
 void
 vdbe_field_ref_create(struct vdbe_field_ref *ref, uint32_t capacity);
 
-/** Check if SQL_EXPR func has single arg. */
-bool
-func_sql_expr_has_single_arg(const struct func *base);
-
 #if defined(__cplusplus)
 } /* extern "C" { */
 #endif
diff --git a/src/box/sql/func.h b/src/box/sql/func.h
index 374a5d83d..f36b6b513 100644
--- a/src/box/sql/func.h
+++ b/src/box/sql/func.h
@@ -16,6 +16,10 @@ struct func_def;
 struct func *
 func_sql_expr_new(const struct func_def *def);
 
+/** Check if SQL_EXPR func has single arg. */
+bool
+func_sql_expr_has_single_arg(const struct func *base);
+
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined __cplusplus */
diff --git a/src/box/tuple_constraint_func.c b/src/box/tuple_constraint_func.c
index 6cd7cba9a..996dc928d 100644
--- a/src/box/tuple_constraint_func.c
+++ b/src/box/tuple_constraint_func.c
@@ -42,8 +42,7 @@ tuple_constraint_func_find(struct tuple_constraint *constr)
  * used for better error message).
  */
 static int
-tuple_constraint_func_verify(struct tuple_constraint *constr,
-			     struct func *func, bool is_field)
+tuple_constraint_func_verify(struct tuple_constraint *constr, struct func *func)
 {
 	const char *func_name = func->def->name;
 
@@ -56,14 +55,6 @@ tuple_constraint_func_verify(struct tuple_constraint *constr,
 				    func_name));
 		return -1;
 	}
-	if (func->def->language == FUNC_LANGUAGE_SQL_EXPR && is_field &&
-	    !func_sql_expr_has_single_arg(func)) {
-		diag_set(ClientError, ER_CREATE_CONSTRAINT, constr->def.name,
-			 constr->space->def->name, "Number of arguments in a "
-			 "SQL field constraint function is greater than one");
-		constr->space = NULL;
-		return -1;
-	}
 	if (!func->def->is_deterministic) {
 		diag_set(ClientError, ER_CREATE_CONSTRAINT,
 			 constr->def.name, constr->space->def->name,
@@ -145,7 +136,7 @@ tuple_constraint_func_unpin(struct tuple_constraint *constr)
 
 int
 tuple_constraint_func_init(struct tuple_constraint *constr,
-			   struct space *space, bool is_field)
+			   struct space *space)
 {
 	assert(constr->def.type == CONSTR_FUNC);
 	constr->space = space;
@@ -158,8 +149,7 @@ tuple_constraint_func_init(struct tuple_constraint *constr,
 		assert(constr->check == tuple_constraint_noop_check);
 		return 0;
 	}
-	if (func == NULL ||
-	    tuple_constraint_func_verify(constr, func, is_field) != 0) {
+	if (func == NULL || tuple_constraint_func_verify(constr, func) != 0) {
 		constr->space = NULL;
 		return -1;
 	}
diff --git a/src/box/tuple_constraint_func.h b/src/box/tuple_constraint_func.h
index ff65f0865..291c3a4e7 100644
--- a/src/box/tuple_constraint_func.h
+++ b/src/box/tuple_constraint_func.h
@@ -19,7 +19,7 @@ struct space;
  */
 int
 tuple_constraint_func_init(struct tuple_constraint *constr,
-			   struct space *space, bool is_field);
+			   struct space *space);
 
 #ifdef __cplusplus
 } /* extern "C" */
```
</details>

The number of changed lines is exactly the same, but everything is done in a single file where it belongs. I also moved `func_sql_expr_has_single_arg()` to `sql/func.h`. Because you had its definition in `sql/func.c`, but somewhy had it declared in `box/sql.h` which is weird.

Reply:
Hi! Thank you for the review! I'm not sure we should check the validity of the constraint outside of `tuple_constraint_func_verify()`. But I will ask the opinion of the second reviewer.